### DISCUSSION TITLE
How to use `serde_as` with vec of my type?

### DISCUSSION THREAD

**Discussion by Magicloud** [jonasbb/serde_with#395] [Q&A] [answered]
Say
```rust
struct MyType;
impl Display for MyType {};
impl FromStr for MyType {};

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
struct MyStruct {
  #[serde_as(as = "DisplayFromStr")] // How to make this line works?
  field: Vec<MyType>
}
```

Thanks.

**Comment by jonasbb** [comment, marked as answer]
Hi, as written, the code tries to use the `Display` an `FromStr` implementations of `Vec` to perform the serialization. These do not exist, which is why you see the error. What you want is to apply the `DisplayFromStr` conversion on each element of the `Vec`. You write this as:

```rust
#[serde_with::serde_as] // You also need to add this line in front of each struct/enum where you use serde_as
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
struct MyStruct {
    #[serde_as(as = "Vec<DisplayFromStr>")]
    field: Vec<MyType>,
}
```

Other container types (e.g., `Option`, `BTreeMap`, `HashSet`) work the same way, by using the collection type for the field and inside `serde_as`.

If you work with a type you defined (like in your example), you might also be interested in the two derive macros [`SerializeDisplay`](https://docs.rs/serde_with/latest/serde_with/derive.SerializeDisplay.html) and [`DeserializeFromStr`](https://docs.rs/serde_with/latest/serde_with/derive.DeserializeFromStr.html). They can be used instead of the derives from `serde`. This way `MyType` implements the necessary traits you can remove `serde_as` from `MyStruct`.

**Reply by Magicloud** [reply]
Simple like that? Cool!

Thanks for the answer.
